File: src/intellibitz/sted/actions/STEDWindowAction.java
package intellibitz.sted.actions;

import intellibitz.sted.event.*;
import intellibitz.sted.launch.STEDGUI;
import intellibitz.sted.ui.STEDWindow;

import javax.swing.*;
import java.awt.event.*;
import java.util.logging.Logger;

public class STEDWindowAction
        extends AbstractAction
        implements WindowListener,
        KeyListener,
        WindowStateListener,
        WindowFocusListener,
        IStatusEventSource,
        IMessageEventSource {

    private STEDWindow stedWindow;
    protected Logger logger;
    private StatusEvent statusEvent;
    private IStatusListener statusListener;
    private MessageEvent messageEvent;
    private IMessageListener messageListener;

    public STEDWindowAction() {
        super();
        logger = Logger.getLogger(getClass().getName());
        statusEvent = new StatusEvent(this);
        messageEvent = new MessageEvent(this);
    }

    public void fireMessagePosted(String message) {
        messageEvent.setMessage(message);
        messageListener.messagePosted(messageEvent);
    }

    public void fireMessagePosted() {
        messageListener.messagePosted(messageEvent);
    }

    public void addMessageListener(IMessageListener messageListener) {
        this.messageListener = messageListener;
    }

    public void fireStatusPosted(String message) {
        statusEvent.setStatus(message);
        statusListener.statusPosted(statusEvent);
    }

    public void fireStatusPosted() {
        statusListener.statusPosted(statusEvent);
    }

    public void addStatusListener(IStatusListener statusListener) {
        this.statusListener = statusListener;
    }

    public void actionPerformed(ActionEvent e) {

    }

    public STEDWindow getSTEDWindow() {
        if (null == stedWindow) {
            stedWindow = STEDGUI.getSTEDWindow();
        }
        return stedWindow;
    }

    public void setSTEDWindow(STEDWindow stedWindow) {
        this.stedWindow = stedWindow;
    }

    protected void showMessageDialog(String message) {
        fireMessagePosted(message);
    }

    /**
     * Invoked when a window has been opened.
     */
    public void windowOpened(WindowEvent e) {
    }

    /**
     * Invoked when a window is in the process of being closed. The close
     * operation can be overridden at this point.
     */
    public void windowClosing(WindowEvent e) {
    }

    /**
     * Invoked when a window has been closed.
     */
    public void windowClosed(WindowEvent e) {
    }

    /**
     * Invoked when a window is iconified.
     */
    public void windowIconified(WindowEvent e) {
    }

    /**
     * Invoked when a window is de-iconified.
     */
    public void windowDeiconified(WindowEvent e) {
    }

    /**
     * Invoked when a window is activated.
     */
    public void windowActivated(WindowEvent e) {
    }

    /**
     * Invoked when a window is de-activated.
     */
    public void windowDeactivated(WindowEvent e) {
    }

    /**
     * Invoked when a window state is changed.
     *
     * @since 1.4
     */
    public void windowStateChanged(WindowEvent e) {
    }

    /**
     * Invoked when the Window is set to be the focused Window, which means that
     * the Window, or one of its subcomponents, will receive keyboard events.
     *
     * @since 1.4
     */
    public void windowGainedFocus(WindowEvent e) {
    }

    /**
     * Invoked when the Window is no longer the focused Window, which means that
     * keyboard events will no longer be delivered to the Window or any of its
     * subcomponents.
     *
     * @since 1.4
     */
    public void windowLostFocus(WindowEvent e) {
    }

    public void keyTyped(KeyEvent e) {
        //To change body of implemented methods use File | Settings | File Templates.
    }

    public void keyPressed(KeyEvent e) {
        //To change body of implemented methods use File | Settings | File Templates.
    }

    public void keyReleased(KeyEvent e) {
        //To change body of implemented methods use File | Settings | File Templates.
    }
}

File: src/main/intellibitz/sted/actions/STEDWindowAction.kt
package sted.actions

import sted.STEDGUI
import sted.event.*
import sted.ui.STEDWindow
import java.awt.event.*
import java.util.logging.Logger
import javax.swing.AbstractAction

open class STEDWindowAction : AbstractAction(), WindowListener, KeyListener, WindowStateListener, WindowFocusListener,
    IStatusEventSource, IMessageEventSource {
    private val statusEvent: StatusEvent
    private val messageEvent: MessageEvent
    //    private var stedWindow: STEDWindow? = null
    val stedWindow: STEDWindow by lazy {
        STEDGUI.sTEDWindow
    }

    @JvmField
    protected var logger = Logger.getLogger(javaClass.name)
    private var statusListener: IStatusListener? = null
    private var messageListener: IMessageListener? = null
    fun fireMessagePosted(message: String?) {
        messageEvent.message = message
        messageListener!!.messagePosted(messageEvent)
    }

    override fun fireMessagePosted() {
        messageListener!!.messagePosted(messageEvent)
    }

    override fun addMessageListener(messageListener: IMessageListener) {
        this.messageListener = messageListener
    }

    fun fireStatusPosted(message: String?) {
        statusEvent.status = message
        statusListener!!.statusPosted(statusEvent)
    }

    override fun fireStatusPosted() {
        statusListener!!.statusPosted(statusEvent)
    }

    override fun addStatusListener(statusListener: IStatusListener) {
        this.statusListener = statusListener
    }

    override fun actionPerformed(e: ActionEvent) {}

    /**
     * Invoked when a window has been opened.
     */
    override fun windowOpened(e: WindowEvent) {}

    /**
     * Invoked when a window is in the process of being closed. The close
     * operation can be overridden at this point.
     */
    override fun windowClosing(e: WindowEvent) {}

    /**
     * Invoked when a window has been closed.
     */
    override fun windowClosed(e: WindowEvent) {}

    /**
     * Invoked when a window is iconified.
     */
    override fun windowIconified(e: WindowEvent) {}

    /**
     * Invoked when a window is de-iconified.
     */
    override fun windowDeiconified(e: WindowEvent) {}

    /**
     * Invoked when a window is activated.
     */
    override fun windowActivated(e: WindowEvent) {}

    /**
     * Invoked when a window is de-activated.
     */
    override fun windowDeactivated(e: WindowEvent) {}

    /**
     * Invoked when a window state is changed.
     *
     * @since 1.4
     */
    override fun windowStateChanged(e: WindowEvent) {}

    /**
     * Invoked when the Window is set to be the focused Window, which means that
     * the Window, or one of its subcomponents, will receive keyboard events.
     *
     * @since 1.4
     */
    override fun windowGainedFocus(e: WindowEvent) {}

    /**
     * Invoked when the Window is no longer the focused Window, which means that
     * keyboard events will no longer be delivered to the Window or any of its
     * subcomponents.
     *
     * @since 1.4
     */
    override fun windowLostFocus(e: WindowEvent) {}
    override fun keyTyped(e: KeyEvent) {
        //To change body of implemented methods use File | Settings | File Templates.
    }

    override fun keyPressed(e: KeyEvent) {
        //To change body of implemented methods use File | Settings | File Templates.
    }

    override fun keyReleased(e: KeyEvent) {
        //To change body of implemented methods use File | Settings | File Templates.
    }

    init {
        statusEvent = StatusEvent(this)
        messageEvent = MessageEvent(this)
    }
}